src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java

```diff
@@ -377,120 +378,155 @@ private String help(MetricRule rule, String unit, String statistic) {
-        request.setNamespace(rule.awsNamespace);
-        request.setMetricName(rule.awsMetricName);
-        request.setStatistics(rule.awsStatistics);
-        request.setExtendedStatistics(rule.awsExtendedStatistics);
-        request.setEndTime(startDate);
-        request.setStartTime(endDate);
-        request.setPeriod(rule.periodSeconds);
-
-        String baseName = safeName(rule.awsNamespace.toLowerCase() + "_" + toSnakeCase(rule.awsMetricName));
-        String jobName = safeName(rule.awsNamespace.toLowerCase());
-        List<MetricFamilySamples.Sample> sumSamples = new ArrayList<MetricFamilySamples.Sample>();
-        List<MetricFamilySamples.Sample> sampleCountSamples = new ArrayList<MetricFamilySamples.Sample>();
-        List<MetricFamilySamples.Sample> minimumSamples = new ArrayList<MetricFamilySamples.Sample>();
-        List<MetricFamilySamples.Sample> maximumSamples = new ArrayList<MetricFamilySamples.Sample>();
-        List<MetricFamilySamples.Sample> averageSamples = new ArrayList<MetricFamilySamples.Sample>();
-        HashMap<String, ArrayList<MetricFamilySamples.Sample>> extendedSamples = new HashMap<String, ArrayList<MetricFamilySamples.Sample>>();
-
-        String unit = null;
-
-        if (rule.awsNamespace.equals("AWS/DynamoDB")
-                && rule.awsDimensions.contains("GlobalSecondaryIndexName")
-                && brokenDynamoMetrics.contains(rule.awsMetricName)) {
-            baseName += "_index";
+    private static String queryId(MetricRule rule, List<Dimension> dimensions, String stat) {
+      String idComponents = new StringBuilder()
+              .append(rule.awsNamespace)
+              .append(rule.awsMetricName)
+              .append(dimensions.hashCode())
+              .append(stat)
+              .toString();
+      // We cant't use the concatenated components as the ID directly because AWS will not accept special chars as ID
+      // Also, the ID must start with a letter, not a digit
+      return "q" + Integer.toUnsignedString(idComponents.hashCode());
+    }
+
+    private static MetricDataQuery metricDataQuery(MetricRule rule, List<Dimension> dimensions, String stat) {
+      Metric metric = new Metric()
+              .withMetricName(rule.awsMetricName)
+              .withNamespace(rule.awsNamespace)
+              .withDimensions(dimensions);
+
+      MetricStat metricStat = new MetricStat()
+              .withMetric(metric)
+              .withPeriod(rule.periodSeconds)
+              .withStat(stat);
+
+      return new MetricDataQuery()
+              .withId(queryId(rule, dimensions, stat))
+              .withMetricStat(metricStat);
+    }
+
+    private static List<MetricDataQuery> queriesForRuleDimensions(MetricRule rule, List<Dimension> dimensions) {
+      List<MetricDataQuery> queries = new ArrayList<MetricDataQuery>();
+
+      for (String stat : rule.allStats()) {
+        queries.add(metricDataQuery(rule, dimensions, stat));
```

Comment:
Presently we get all the stats as one request, could this blow up costs by up to 5x if all stats are requested?

Reply:
No, I verified with AWS support and tested. Cost stays the same.

Reply:
Can you add a comment to that effect, and update the README?